Bluetooth: Mesh: Modularizing the proxy

LingaoM opened the pull request:
The Bluetooth proxy feature includes proxy client and proxy server.
In addition to the proxy pdu message used above, pb-gatt also uses
the same proxy pdu message.

Currently zephyr bluetooth mesh couples them in one file.

A file at the separation is called gatt_services.c,
which is used to contain Mesh Provisioning Service and Mesh Proxy Service.
Another file in the separation is called proxy_msg.c,
which is used to process Proxy pdu messages.

Also according to Trond's suggestion:

Rename `CONFIG_BT_MESH_PROXY` to `CONFIG_BT_MESH_GATT`.
Create an additional promptless entry `CONFIG_BT_MESH_GATT_SERVER`
that selects `CONFIG_BT_MESH_GATT` and is selected by
`CONFIG_BT_MESH_GATT_PROXY` or `CONFIG_BT_MESH_PB_GATT`.

Create additional `CONFIG_BT_MESH_PROXY` used to represent proxy feature
(also include proxy client).

see https://github.com/zephyrproject-rtos/zephyr/issues/36343

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik commented on subsys/bluetooth/mesh/proxy.h:
```diff
@@ -14,11 +14,47 @@
+
 int bt_mesh_proxy_prov_enable(void);
 int bt_mesh_proxy_prov_disable(bool disconnect);
 
```
This is a bit of a hassle, perhaps, but ideally, this header should be split up according to the new source file split, and the functions be named appropriately

LingaoM replied:
The file has been split, but it's better to change the name in a PR with the split services.

trond-snekvik commented on subsys/bluetooth/mesh/proxy_msg.c:
```diff
@@ -0,0 +1,249 @@
+/*  Bluetooth Mesh */
```
Would "gatt_msg.c" be more appropriate? I know this is just following the original suggestion though :)

LingaoM replied:
Should we follow the definition in spec?

LingaoM replied:
What does the definition in the specification mean to me is that the pb-gatt message borrows the definition of the proxy format? So this is my point of view, may I ask @jhedberg  point of view?

trond-snekvik commented on subsys/bluetooth/mesh/proxy.h:
```diff
@@ -14,11 +14,47 @@
+
+	struct {
+		proxy_send_cb_t send_cb;
+		proxy_recv_cb_t recv_cb;
```
The `_cb` suffix seems kind of redundant

trond-snekvik commented on subsys/bluetooth/mesh/proxy_msg.c:
```diff
@@ -0,0 +1,249 @@
+		return;
+	}
+
+	if (role->cb.recv_cb) {
```
This is never false

trond-snekvik reviewed (commented):
#36343 suggests splitting the higher levels into proxy_server and pb_gatt as well, this appears to be missing in this PR. To me, this seems like the module that could benefit the most from a split, as a large portion of what is now gatt_services is gatt proxy specific, and not really relevant for the provisioning part of the subsystem. We're currently using the proxy filter type to distinguish these two roles, which feels like a bit of a hack to me, at least on an architectural level. 

Is this upper level split skipped because it doesn't make sense after all, or is it more about reducing the scope of the PR?

LingaoM commented on subsys/bluetooth/mesh/proxy.h:
```diff
@@ -14,11 +14,47 @@
+
+	struct k_work_delayable sar_timer;
+	struct net_buf_simple buf;
+};
```
ye s

LingaoM commented:
> #36343 suggests splitting the higher levels into proxy_server and pb_gatt as well, this appears to be missing in this PR. To me, this seems like the module that could benefit the most from a split, as a large portion of what is now gatt_services is gatt proxy specific, and not really relevant for the provisioning part of the subsystem. We're currently using the proxy filter type to distinguish these two roles, which feels like a bit of a hack to me, at least on an architectural level. 
> 
> 
> 
> Is this upper level split skipped because it doesn't make sense after all, or is it more about reducing the scope of the PR?

I think if we want to split gatt_sevices.c, then this file basically has nothing.

trond-snekvik commented:
> I think if we want to split gatt_sevices.c, then this file basically has nothing.
That's probably true, yes. In this case, perhaps it shouldn't exist? There are currently 11 instances of `#if defined(CONFIG_BT_MESH_GATT_PROXY)` and `#if defined(CONFIG_BT_MESH_PB_GATT)` in that file, and a lot of functions that are specific to each of the two roles.

I can understand that you don't want to take this on in this PR, which already improves the situation quite a bit. I'm okay with the current scope, and will approve once the open comments are resolved.

LingaoM commented:
> > I think if we want to split gatt_sevices.c, then this file basically has nothing.
> > That's probably true, yes. In this case, perhaps it shouldn't exist? There are currently 11 instances of `#if defined(CONFIG_BT_MESH_GATT_PROXY)` and `#if defined(CONFIG_BT_MESH_PB_GATT)` in that file, and a lot of functions that are specific to each of the two roles.
> 
> I can understand that you don't want to take this on in this PR, which already improves the situation quite a bit. I'm okay with the current scope, and will approve once the open comments are resolved.

I did try to separate services yesterday, but I think it is more appropriate to deal with it in the following PR.

LingaoM commented on subsys/bluetooth/mesh/CMakeLists.txt:
```diff
@@ -41,7 +41,11 @@ zephyr_library_sources_ifdef(CONFIG_BT_MESH_PB_ADV pb_adv.c)
-zephyr_library_sources_ifdef(CONFIG_BT_MESH_PROXY proxy.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MESH_GATT_SERVER gatt_services.c)
+
+if (CONFIG_BT_MESH_PROXY OR CONFIG_BT_MESH_PB_GATT)
```
Should we follow the definition in spec?

LingaoM commented on subsys/bluetooth/mesh/CMakeLists.txt:
```diff
@@ -41,7 +41,11 @@ zephyr_library_sources_ifdef(CONFIG_BT_MESH_PB_ADV pb_adv.c)
+if (CONFIG_BT_MESH_GATT_SERVER OR CONFIG_BT_MESH_GATT)
+	zephyr_library_sources(proxy_msg.c)
+endif()
 
```
Need to make it clear, CONFIG_BT_MESH_GATT is most suitable here. This is because this is selected by pb-gatt-srv and proxy-srv, and the future pb-gatt-cli and proxy-cli are selected. @trond-snekvik Am I right?

trond-snekvik replied:
I agree :+1:

LingaoM commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -101,20 +101,29 @@ endif # BT_MESH_CDB
+	bool
 	select BT_GATT_DYNAMIC_DB
 
+config BT_MESH_GATT_SERVER
```
select BT_GATT_DYNAMIC_DB should place here, not above.

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -174,7 +174,7 @@ static void send_pending_adv(struct k_work *work)
 
 	/* No more pending buffers */
-	if (IS_ENABLED(CONFIG_BT_MESH_PROXY)) {
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT)) {
```
Also think should replace by CONFIG_BT_MESH_GATT_SERVER

LingaoM replied:
Since it only selected by proxy server or pb gatt server.

LingaoM commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -108,7 +108,7 @@ static void adv_thread(void *p1, void *p2, void *p3)
 		struct net_buf *buf;
 
-		if (IS_ENABLED(CONFIG_BT_MESH_PROXY)) {
+		if (IS_ENABLED(CONFIG_BT_MESH_GATT)) {
```
Also think should replace by CONFIG_BT_MESH_GATT_SERVER

LingaoM commented on subsys/bluetooth/mesh/crypto.c:
```diff
@@ -292,7 +292,7 @@ int bt_mesh_net_encrypt(const uint8_t key[16], struct net_buf_simple *buf,
 	BT_DBG("PDU (len %u) %s", buf->len, bt_hex(buf->data, buf->len));
 
-	if (IS_ENABLED(CONFIG_BT_MESH_PROXY) && proxy) {
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT) && proxy) {
```
This should not be changed, because CONFIG_BT_MESH_PROXY is used by CONFIG_BT_MESH_GATT_PROXY and CONFIG_BT_MESH_PROXY_CLIENT(in the future).

LingaoM commented on subsys/bluetooth/mesh/crypto.c:
```diff
@@ -319,7 +319,7 @@ int bt_mesh_net_decrypt(const uint8_t key[16], struct net_buf_simple *buf,
 	       mic_len);
 
-	if (IS_ENABLED(CONFIG_BT_MESH_PROXY) && proxy) {
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT) && proxy) {
```
Same above too.

LingaoM commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -141,23 +150,23 @@ config BT_MESH_NODE_ID_TIMEOUT
 	default 1
 	range 1 32767
-	depends on BT_MESH_PROXY
+	depends on BT_MESH_GATT_SERVER
```
Also should be depends on BT_MESH_GATT_PROXY by ideally.
But now being coupled by pb-gatt is really annoying, but I will solve this problem completely in the next pr.

LingaoM commented on include/linker/common-rom.ld:
```diff
@@ -132,7 +132,7 @@
 #endif
 
-#if defined(CONFIG_BT_MESH_PROXY)
+#if defined(CONFIG_BT_MESH_GATT)
```
Also should be CONFIG_BT_MESH_GATT_PROXY!

LingaoM commented:
@trond-snekvik Please review my view above, thank you very much.

trond-snekvik commented:
> @trond-snekvik Please review my view above, thank you very much.

All comments make sense to me :+1: 

I'm afraid I created a conflict for you in proxy.c in #36513, sorry about that.

LingaoM commented:
> > @trond-snekvik Please review my view above, thank you very much.
> 
> All comments make sense to me 👍
> 
> I'm afraid I created a conflict for you in proxy.c in #36513, sorry about that.

Don't worry, I will update.

LingaoM committed:
Bluetooth: Mesh: Modularizing the proxy

The Bluetooth proxy feature includes proxy client
and proxy server. In addition to the proxy pdu
message used above, pb-gatt also uses the same
proxy pdu message.

Currently zephyr bluetooth mesh couples them in one file.

A file at the separation is called gatt_services.c,
which is used to contain Mesh Provisioning Service
and Mesh Proxy Service.
Another file in the separation is called proxy_msg.c,
which is used to process Proxy pdu messages.

Also according to Trond's suggestion:

Rename `CONFIG_BT_MESH_PROXY` to `CONFIG_BT_MESH_GATT`.
Create an additional promptless entry
`CONFIG_BT_MESH_GATT_SERVER` that selects
`CONFIG_BT_MESH_GATT` and is selected by
`CONFIG_BT_MESH_GATT_PROXY` or `CONFIG_BT_MESH_PB_GATT`.

Create additional `CONFIG_BT_MESH_PROXY` used to represent
proxy feature (also include proxy client).

see #36343

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Rename proxy.c to proxy_msg.c

The Bluetooth proxy feature includes proxy client
and proxy server. In addition to the proxy pdu
message used above, pb-gatt also uses the same
proxy pdu message.

Currently zephyr bluetooth mesh couples them in one file.

A file at the separation is called gatt_services.c,
which is used to contain Mesh Provisioning Service and
Mesh Proxy Service.

Another file in the separation is called proxy_msg.c,
which is used to process Proxy pdu messages.

Also according to Trond's suggestion:

Rename CONFIG_BT_MESH_PROXY to CONFIG_BT_MESH_GATT.
Create an additional promptless entry CONFIG_BT_MESH_GATT_SERVER
that selects CONFIG_BT_MESH_GATT and is selected by
CONFIG_BT_MESH_GATT_PROXY or CONFIG_BT_MESH_PB_GATT.

Create additional CONFIG_BT_MESH_PROXY used to represent
proxy feature (also include proxy client).

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Split proxy.h to proxy.h and proxy_msg.h

The Bluetooth proxy feature includes proxy client and proxy server.
In addition to the proxy pdu message used above, pb-gatt also uses
the same proxy pdu message.

Currently zephyr bluetooth mesh couples them in one file.

A file at the separation is called gatt_services.c,
which is used to contain Mesh Provisioning Service and
Mesh Proxy Service.

Another file in the separation is called proxy_msg.c,
which is used to process Proxy pdu messages.

Also according to Trond's suggestion:

Rename CONFIG_BT_MESH_PROXY to CONFIG_BT_MESH_GATT.
Create an additional promptless entry CONFIG_BT_MESH_GATT_SERVER
that selects CONFIG_BT_MESH_GATT and is selected by
CONFIG_BT_MESH_GATT_PROXY or CONFIG_BT_MESH_PB_GATT.

Create additional CONFIG_BT_MESH_PROXY used to represent
proxy feature (also include proxy client).

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

trond-snekvik reviewed (approved):
LGTM.

LingaoM commented:
@jhedberg Please review, thanks.
